lib/ClangImporter/SwiftDeclSynthesizer.cpp

```diff
@@ -2534,128 +2534,160 @@ SwiftDeclSynthesizer::makeDefaultArgument(const clang::ParmVarDecl *param,
 
 // MARK: C++ foreign reference type constructors
 
-clang::CXXMethodDecl *
+llvm::SmallVector<clang::CXXMethodDecl *, 4>
 SwiftDeclSynthesizer::synthesizeStaticFactoryForCXXForeignRef(
     const clang::CXXRecordDecl *cxxRecordDecl) {
 
   clang::ASTContext &clangCtx = cxxRecordDecl->getASTContext();
   clang::Sema &clangSema = ImporterImpl.getClangSema();
 
   clang::QualType cxxRecordTy = clangCtx.getRecordType(cxxRecordDecl);
+  clang::SourceLocation cxxRecordDeclLoc = cxxRecordDecl->getLocation();
 
-  clang::CXXConstructorDecl *defaultCtorDecl = nullptr;
-  for (clang::CXXConstructorDecl *ctor : cxxRecordDecl->ctors()) {
-    if (ctor->parameters().empty() && !ctor->isDeleted() &&
-        ctor->getAccess() != clang::AS_private &&
-        ctor->getAccess() != clang::AS_protected) {
-      defaultCtorDecl = ctor;
-      break;
-    }
-  }
-  if (!defaultCtorDecl)
-    return nullptr;
+  llvm::SmallVector<clang::CXXMethodDecl *, 4> synthesizedFactories;
 
   clang::FunctionDecl *operatorNew = nullptr;
   clang::FunctionDecl *operatorDelete = nullptr;
   bool passAlignment = false;
   bool findingAllocFuncFailed = clangSema.FindAllocationFunctions(
-      cxxRecordDecl->getLocation(), clang::SourceRange(), clang::Sema::AFS_Both,
-      clang::Sema::AFS_Both, cxxRecordTy,
-      /*IsArray*/ false, passAlignment, clang::MultiExprArg(), operatorNew,
-      operatorDelete, /*Diagnose*/ false);
+      cxxRecordDeclLoc, clang::SourceRange(), clang::Sema::AFS_Both,
+      clang::Sema::AFS_Both, cxxRecordTy, /*IsArray=*/false, passAlignment,
+      clang::MultiExprArg(), operatorNew, operatorDelete, /*Diagnose=*/false);
   if (findingAllocFuncFailed || !operatorNew || operatorNew->isDeleted() ||
       operatorNew->getAccess() == clang::AS_private ||
       operatorNew->getAccess() == clang::AS_protected)
-    return nullptr;
+    return synthesizedFactories;
```

Comment:
nit: perhaps change this to `return {};` or something to imply that we are intentionally returning an empty list because this is an exceptional case?